codex-rs/core/src/codex.rs

```diff
@@ -193,6 +193,90 @@ impl Session {
+/// *never* used for billing – only for displaying approximate usage stats to
+/// the user.
+fn approx_token_count(s: &str) -> usize {
+    // Avoid division by zero for empty strings.
+    if s.is_empty() {
+        0
+    } else {
+        (s.len() + 3) / 4 // round up
+    }
+}
+
+/// Counts the number of tokens contained in a collection of [`ResponseItem`]s
+/// by summing up the textual content of all `InputText` and `OutputText`
+/// elements.
+fn count_tokens_in_items(items: &[ResponseItem]) -> usize {
+    items
+        .iter()
+        .map(|item| match item {
+            ResponseItem::Message { content, .. } => content
+                .iter()
+                .filter_map(|c| match c {
+                    ContentItem::InputText { text } | ContentItem::OutputText { text } => {
+                        Some(approx_token_count(text))
+                    }
+                    _ => None,
+                })
+                .sum::<usize>(),
+            _ => 0,
+        })
+        .sum()
+}
+
+/// Returns the OpenAI per-1K-token pricing (prompt, completion) **in USD** for
+/// a given model name. The list is *not* exhaustive – it only covers the most
+/// common public models so we offer reasonable estimates without hard-coding
+/// every single variant. Unknown models default to `None` so callers can fall
+/// back gracefully.
+fn get_openai_pricing(model: &str) -> Option<(f64, f64)> {
+    // Exact mapping (per *token* rates, not per-1K)
+    let detailed: &[(&str, (f64, f64))] = &[ // (model, (input, output))
+        ("o3", (10.0 / 1_000_000.0, 40.0 / 1_000_000.0)),
+        ("o4-mini", (1.1 / 1_000_000.0, 4.4 / 1_000_000.0)),
+        ("gpt-4.1-nano", (0.1 / 1_000_000.0, 0.4 / 1_000_000.0)),
+        ("gpt-4.1-mini", (0.4 / 1_000_000.0, 1.6 / 1_000_000.0)),
+        ("gpt-4.1", (2.0 / 1_000_000.0, 8.0 / 1_000_000.0)),
+        ("gpt-4o-mini", (0.6 / 1_000_000.0, 2.4 / 1_000_000.0)),
+        ("gpt-4o", (5.0 / 1_000_000.0, 20.0 / 1_000_000.0)),
+        ("codex-mini-latest", (1.5 / 1_000_000.0, 6.0 / 1_000_000.0)),
+    ];
+
+    let key = model.to_ascii_lowercase();
+    if let Some((in_rate, out_rate)) = detailed
+        .iter()
+        .find(|(m, _)| key.starts_with(*m))
+        .map(|(_, r)| *r)
+    {
+        return Some((in_rate, out_rate));
+    }
+
+    // Fallback coarse buckets (per-1K rates → convert to per-token)
+    let per_1k_to_per_token = |x: f64| x / 1000.0;
+    if key.contains("gpt-4o") {
+        return Some((per_1k_to_per_token(0.005), per_1k_to_per_token(0.015)));
+    }
+    if key.contains("gpt-4-turbo") {
+        return Some((per_1k_to_per_token(0.01), per_1k_to_per_token(0.03)));
+    }
+    if key.contains("gpt-4") {
+        return Some((per_1k_to_per_token(0.03), per_1k_to_per_token(0.06)));
+    }
+    if key.contains("gpt-3.5-turbo") {
+        return Some((per_1k_to_per_token(0.0005), per_1k_to_per_token(0.0015)));
+    }
+    None
```

Comment:
If we're not certain about the cost let's not display it at all? Feels a bit hacky to use per~1k rates like this